// react-annotate-text/dist/annotation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Annotaion(_ref) {
  var highlightData = _ref.highlightData,
      scrollPosition = _ref.scrollPosition,
      hoverChange = _ref.hoverChange;

  var _onMouseOver = function _onMouseOver(id) {
    hoverChange(id);
  };

  var findAnnotations = function findAnnotations() {
    return highlightData.map(function (singleAnnotation) {
      return singleAnnotation.position.map(function (rectangle, index) {
        return _react2.default.createElement(
          "div",
          { key: singleAnnotation.id + "_" + index },
          _react2.default.createElement("div", {
            className: "rectangle-not-visible",
            onMouseOver: function onMouseOver() {
              return _onMouseOver(singleAnnotation.id);
            },
            id: singleAnnotation.id,
            style: {
              position: "absolute",
              height: rectangle.height,
              top: rectangle.top - scrollPosition.scrollY,
              left: rectangle.left - scrollPosition.scrollX,
              width: rectangle.width,
              backgroundColor: "none",
              opacity: 1
            }
          }),
          _react2.default.createElement("div", {
            className: "rectangle-visible",
            id: singleAnnotation.id,
            style: {
              position: "absolute",
              height: rectangle.height,
              top: rectangle.top - scrollPosition.scrollY,
              left: rectangle.left - scrollPosition.scrollX,
              width: rectangle.width,
              backgroundColor: "yellow",
              opacity: 1,
              zIndex: -1
            }
          })
        );
      });
    });
  };

  return _react2.default.createElement(
    "div",
    { className: "annotations" },
    findAnnotations()
  );
}
exports.default = Annotaion;
